# Question1: Predicting Heart Disease

In [2]:
import numpy as np
import pandas as pd
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier

from sklearn.preprocessing import OneHotEncoder

In [3]:
df = pd.read_csv("https://github.com/mpourhoma/CS4661/raw/master/Heart_s.csv")

In [4]:
feature_cols = ['Age','RestBP','Chol','RestECG','MaxHR','Oldpeak']

X = df[feature_cols] 
print(X.shape)

(301, 6)


In [5]:
y = df['AHD']
print(y.shape)

(301,)


In [6]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=6)

## KNN

In [7]:
k = 3
hd_knn = KNeighborsClassifier(n_neighbors=k)

hd_knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [8]:
y_predict = hd_knn.predict(X_test)

In [9]:
accuracy = accuracy_score(y_test, y_predict)

print(accuracy)

0.6447368421052632


## Decision Tree

In [10]:
X_train2, X_test2, y_train2, y_test2 = train_test_split(X, y, test_size=0.25)

In [11]:
my_decisiontree = DecisionTreeClassifier(random_state=5)

In [12]:
my_decisiontree.fit(X_train2, y_train2)

DecisionTreeClassifier(random_state=5)

In [13]:
y_predict_dt = my_decisiontree.predict(X_test2)

In [14]:
score_dt = accuracy_score(y_test2, y_predict_dt)

In [15]:
print(score_dt)

0.6052631578947368


## Logistic Regression Classifiers

In [21]:
my_logreg = LogisticRegression(max_iter=301)

In [22]:
my_logreg.fit(X_train, y_train)

LogisticRegression()

In [23]:
y_predict_lr = my_logreg.predict(X_test)

In [24]:
score_lr = accuracy_score(y_test, y_predict_lr)

In [25]:
print(score_lr)

0.6710526315789473


### Linear Regression had the highest accuracy at 67.105%

## Categorical Features

### OneHotEncoding

In [11]:
ohe = OneHotEncoder()

In [23]:
feature_array = ohe.fit_transform(df[["Gender", "ChestPain", "Thal"]]).toarray()

In [21]:
feature_labels = ['f','m','asymptomatic', 'nonanginal', 'nontypical', 'typical','fixed', 'normal', 'reversable']
print(feature_labels)

['f', 'm', 'asymptomatic', 'nonanginal', 'nontypical', 'typical', 'fixed', 'normal', 'reversable']


In [24]:
features = pd.DataFrame(feature_array, columns=feature_labels)
print(features)

       f    m  asymptomatic  nonanginal  nontypical  typical  fixed  normal  \
0    1.0  0.0           0.0         0.0         0.0      1.0    1.0     0.0   
1    1.0  0.0           1.0         0.0         0.0      0.0    0.0     1.0   
2    1.0  0.0           1.0         0.0         0.0      0.0    0.0     0.0   
3    1.0  0.0           0.0         1.0         0.0      0.0    0.0     1.0   
4    0.0  1.0           0.0         0.0         1.0      0.0    0.0     1.0   
..   ...  ...           ...         ...         ...      ...    ...     ...   
296  1.0  0.0           0.0         0.0         0.0      1.0    0.0     0.0   
297  1.0  0.0           1.0         0.0         0.0      0.0    0.0     0.0   
298  1.0  0.0           1.0         0.0         0.0      0.0    0.0     0.0   
299  0.0  1.0           0.0         0.0         1.0      0.0    0.0     1.0   
300  1.0  0.0           0.0         1.0         0.0      0.0    0.0     1.0   

     reversable  
0           0.0  
1           0.0

In [26]:
df_new = pd.concat([df, features], axis=1)
df_new.head()

Age Gender     ChestPain  RestBP  Chol  RestECG  MaxHR  Oldpeak  \
0   63      f       typical     145   233        2    150      2.3   
1   67      f  asymptomatic     160   286        2    108      1.5   
2   67      f  asymptomatic     120   229        2    129      2.6   
3   37      f    nonanginal     130   250        0    187      3.5   
4   41      m    nontypical     130   204        2    172      1.4   

         Thal  AHD    f    m  asymptomatic  nonanginal  nontypical  typical  \
0       fixed   No  1.0  0.0           0.0         0.0         0.0      1.0   
1      normal  Yes  1.0  0.0           1.0         0.0         0.0      0.0   
2  reversable  Yes  1.0  0.0           1.0         0.0         0.0      0.0   
3      normal   No  1.0  0.0           0.0         1.0         0.0      0.0   
4      normal   No  0.0  1.0           0.0         0.0         1.0      0.0   

   fixed  normal  reversable  
0    1.0     0.0         0.0  
1    0.0     1.0         0.0  
2    0.0     0.0         1.0  
3    0.0     1.0         0.0  
4    0.0     1.0         0.0

### Repeat split with new OHE columns

In [27]:
feature_cols = ['Age','RestBP','Chol','RestECG','MaxHR','Oldpeak', 'f','m','asymptomatic', 'nonanginal', 'nontypical', 'typical','fixed', 'normal', 'reversable']

X = df_new[feature_cols] 
print(X.shape)

(301, 15)


In [28]:
y = df['AHD']
print(y.shape)

(301,)


In [29]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=6)

### Repeat KNN

In [30]:
k = 3
hd_knn = KNeighborsClassifier(n_neighbors=k)

hd_knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [31]:
y_predict = hd_knn.predict(X_test)

In [32]:
accuracy = accuracy_score(y_test, y_predict)

print(accuracy)

0.6447368421052632


### Repeat Decision Tree

In [33]:
X_train2, X_test2, y_train2, y_test2 = train_test_split(X, y, test_size=0.25, random_state=5)

In [34]:
my_decisiontree = DecisionTreeClassifier()

In [35]:
my_decisiontree.fit(X_train2, y_train2)

DecisionTreeClassifier()

In [36]:
y_predict_dt = my_decisiontree.predict(X_test2)

In [37]:
score_dt = accuracy_score(y_test2, y_predict_dt)

In [38]:
print(score_dt)

0.8157894736842105


### Repeat Logistic Regression Classifiers

In [76]:
my_logreg = LogisticRegression(max_iter=2000)

In [77]:
my_logreg.fit(X_train, y_train)

LogisticRegression(max_iter=2000)

In [78]:
y_predict_lr = my_logreg.predict(X_test)

In [79]:
score_lr = accuracy_score(y_test, y_predict_lr)

In [80]:
print(score_lr)

0.7763157894736842
